### PR TITLE
Close #LGVISIUM-85: Not found features do now return a 404

### PR DESCRIPTION
**Pull request by dcleres**
Error+Detail instead of Success+Detail for extract_data if no data could be read out

### CODE REVIEW

**Review by stijnvermeeren-swisstopo — CHANGES_REQUESTED** [review]
How do we get these error responses to be properly documented in the Swagger / OpenAPI docs? At the moment, I only see documentation for the success response and for a 422 validation error (which I'm not sure we even use?).

image

File: src/app/api/v1/endpoints/extract_data.py [hunk 1]
```diff
@@ -189,11 +192,14 @@ def extract_text(pdf_page: fitz.Page, user_defined_bbox: fitz.Rect) -> ExtractDa
         text += text_line.text + " "
         text_based_bbox = text_based_bbox + text_line.rect
 
-    bbox = BoundingBox.load_from_fitz_rect(text_based_bbox)
-    return ExtractTextResponse(bbox=bbox, text=text)
+    if text:
+        bbox = BoundingBox.load_from_fitz_rect(text_based_bbox)
+        return ExtractTextResponse(bbox=bbox, text=text)
+    else:
+        return HTTPException(status_code=404, detail="Text not found.")
```
[the remark this answers is not in the document]
Comment (dcleres) [reply src/app/api/v1/endpoints/extract_data.py:199]: Thank you for catching this. I was also not super happy about the way the exceptions were handled when I worked on it. I will try again today to clean this up and see if, with fresh eyes, I am able to find a nicer solution.

**Comment by dcleres** [comment]
@stijnvermeeren-swisstopo Thank you very much for the good code review. FastAPI is creating the 422 by default. I had to disable it manually by changing the OpenAPI Settings. Now the error code should only be the one that we can actually trigger. 

I also cleaned up the code regarding how we raise the different validation errors.

File: src/app/main.py [hunk 2]
```diff
@@ -22,6 +23,33 @@
 app = FastAPI(root_path=f"/{root_path}")
 
 
+def custom_openapi():
+    """Custom function to modify the OpenAPI schema and remove 422 errors."""
```
Comment (stijnvermeeren-swisstopo) [inline src/app/main.py:27]: Please add a link to https://github.com/fastapi/fastapi/discussions/6695, which is where this implementation seems to come from, and is also where additional context can be found (as without additional context, this code looks really hacky and suspicious).

File: src/app/api/v1/endpoints/extract_data.py [hunk 3]
```diff
@@ -168,7 +171,7 @@ def create_response(coord, srs):
     return None
 
 
-def extract_text(pdf_page: fitz.Page, user_defined_bbox: fitz.Rect) -> ExtractDataResponse:
+def extract_text(pdf_page: fitz.Page, user_defined_bbox: fitz.Rect) -> ExtractDataResponse | HTTPException:
```
Comment (stijnvermeeren-swisstopo) [inline src/app/api/v1/endpoints/extract_data.py:174]: Can be reverted to just `ExtractDataResponse` as return type? Also, can't we be more specific and put `ExtractTextResponse` as return type?

File: src/app/api/v1/endpoints/extract_data.py [hunk 4]
```diff
@@ -106,6 +106,9 @@ def extract_data(extract_data_request: ExtractDataRequest) -> ExtractDataRespons
     elif extract_data_request.format == FormatTypes.NUMBER:
         extracted_number = extract_number(pdf_page, user_defined_bbox.to_fitz_rect())
 
+        if isinstance(extracted_number, HTTPException):
+            return extracted_number
```
Comment (stijnvermeeren-swisstopo) [inline src/app/api/v1/endpoints/extract_data.py:110]: No longer needed?

File: src/app/api/v1/endpoints/extract_data.py [hunk 5]
```diff
@@ -93,6 +90,9 @@ def extract_data(extract_data_request: ExtractDataRequest) -> ExtractDataRespons
     elif extract_data_request.format == FormatTypes.TEXT:
         extracted_text: ExtractTextResponse = extract_text(pdf_page, user_defined_bbox.to_fitz_rect())
 
+        if isinstance(extracted_text, HTTPException):
+            return extracted_text
```
Comment (stijnvermeeren-swisstopo) [inline src/app/api/v1/endpoints/extract_data.py:94]: No longer needed?

File: src/app/api/v1/endpoints/extract_data.py [hunk 6]
```diff
@@ -189,11 +192,14 @@ def extract_text(pdf_page: fitz.Page, user_defined_bbox: fitz.Rect) -> ExtractDa
...
-    return ExtractTextResponse(bbox=bbox, text=text)
+    if text:
+        bbox = BoundingBox.load_from_fitz_rect(text_based_bbox)
+        return ExtractTextResponse(bbox=bbox, text=text)
+    else:
+        raise HTTPException(status_code=404, detail="Text not found.")
 
 
-def extract_number(pdf_page: fitz.Page, user_defined_bbox: fitz.Rect) -> ExtractNumberResponse:
+def extract_number(pdf_page: fitz.Page, user_defined_bbox: fitz.Rect) -> ExtractNumberResponse | HTTPException:
```
Comment (stijnvermeeren-swisstopo) [inline src/app/api/v1/endpoints/extract_data.py:202]: Can be reverted to just `ExtractNumberResponse` as return type?

**Review by stijnvermeeren-swisstopo — CHANGES_REQUESTED** [review]
Nice, this looks much cleaner already. Only the removal of the 422 response from the docs looks quite hacky, but as this is the recommended solution on the Github discussion, I'm happy to keep it, but I would just add a link to that discussion. Apart from that, here are just a few small bits that can probably be cleanup up before we merge.

File: src/app/main.py (hunk 2 above)
Comment (dcleres) [reply src/app/main.py:27]: Yes, exactly. This is where I found the fix. I will add it to the code.

File: src/app/api/v1/endpoints/extract_data.py (hunk 4 above)
Comment (dcleres) [reply src/app/api/v1/endpoints/extract_data.py:110]: Indeed. Good catch!

File: src/app/api/v1/endpoints/extract_data.py (hunk 6 above)
Comment (dcleres) [reply src/app/api/v1/endpoints/extract_data.py:202]: Exactly!

**Review by stijnvermeeren-swisstopo — APPROVED** [review]

**Pull request merged**